### ISSUE TITLE
:cpp:enumerator: not linking to the enumeration value

### ISSUE THREAD

**Issue by carlescufi** [michaeljones/breathe#437]
I am using the following snippet:

```
:cpp:enumerator:`NRF_CLOUD_EVT_ERROR` 
```

which should in turn correctly link to the `NRF_CLOUD_EVT_ERROR` entry that is properly generated when using the `.. doxygengroup::` directive. Instead, the generated HTML does not contain a link, although it correctly formats the text as code.

**Comment by melvinvermeeren** [comment]
I am thinking this might be an issue in Sphinx's C++ domain, could also be Breathe does not correctly handle enums though. @jakobandersen anything you are aware of that could be causing this?

Just in case, what versions of breathe/sphinx are you using?

**Comment by jakobandersen** [comment]
It's hard to say without more information. @carlescufi, besides the version of Sphinx and Breathe, can you create or point to a code base with the problem?

**Comment by carlescufi** [comment]
Sure, I used Sphinx 2.0.1 and Breathe 4.13.0.post0.

Perhaps the simplest way to reproduce is using Zephyr upstream:
https://github.com/zephyrproject-rtos/zephyr/tree/master/doc

(note that the official Sphinx and breathe versions for upstream Zephyr are older, but it builds fine with the latest, which I used)

An example enumerator that is not correctly processed can be found in this page:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/guides/bluetooth/bluetooth-arch.rst

For example, :cpp:enumerator:`BT_SECURITY_LOW`, which as you can see is rendered without a link, unlike the functions and types:
https://docs.zephyrproject.org/latest/guides/bluetooth/bluetooth-arch.html#security

**Comment by jakobandersen** [comment]
I think I found the problem:
1. The Sphinx C++ domain has an ``enum`` object type which may be scoped or unscoped.
2. There is also an ``enumerator`` object type. If it occurs inside an unscoped enum, then it is automatically added to the parent scope as well, to get the proper lookup properties.
3. It seems that Zephyr does not contain a documentation entry for ``bt_conn`` which is the parent scope of ``BT_SECURITY_LOW``.
4. When the ``bt_conn::BT_SECURITY_LOW`` enumerator is encountered, a scope for ``bt_conn`` is created without an attached declaration.
5. Therefore, ``BT_SECURITY_LOW`` is not inside an unscoped enum, not even inside an enum, and the adding-to-parent-scope is not being done. (perhaps a warning could/should be issued about enumerators not inside enums?)
6. An unqualified lookup of ``BT_SECURITY_LOW`` gives no result.

The proper fix is to actually document ``bt_conn``, though if this can not easily be done, then a hax may be to qualify the references to the enumerators.

**Comment by carlescufi** [comment]
@jakobandersen 
Thanks so much for digging into this!

to clarify:

>It seems that Zephyr does not contain a documentation entry for `bt_conn` which is the parent scope of `BT_SECURITY_LOW`.

When you say "documentation entry" what do you mean exactly? So I can try and add one to see if it solves the issue.

**Comment by jakobandersen** [comment]
Hmm, I based it on the debug output I got when trying to compile the documentation locally. Though, based on https://docs.zephyrproject.org/latest/reference/bluetooth/connection_mgmt.html the picture gets a bit stranger:
Consider [``BT_SECURITY_LOW``](https://docs.zephyrproject.org/latest/reference/bluetooth/connection_mgmt.html?highlight=bt_security_low#_CPPv3N7bt_conn15BT_SECURITY_LOWE), visually it look like it is scoped inside ``bt_security_t``, but actually Sphinx is getting told that it is inside ``bt_conn`` (which is not how the original code looks?).  Actually, Sphinx is also told ``bt_security_t`` is scoped inside ``bt_conn``.'
That is, Sphinx gets something equivalent to:
```rst
.. cpp:enum:: bt_conn::bt_security_t

.. cpp:enumerator:: bt_conn::BT_SECURITY_LOW
```
I'm not sure where in Doxygen/Breathe the translation goes wrong though.

**Comment by carlescufi** [comment]
@jakobandersen 

>Consider BT_SECURITY_LOW, visually it look like it is scoped inside bt_security_t, but actually Sphinx is getting told that it is inside bt_conn (which is not how the original code looks?)

I think that must be due to the fact that `bt_security_t` is part of the `bt_conn` doxygen group: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/conn.h#L15

**Comment by carlescufi** [comment]
@jakobandersen I verified that the issue is indeed the one you mentioned. If I instead use:

```
:cpp:enumerator:`BT_SECURITY_LOW <bt_conn::BT_SECURITY_LOW>` 
```

It displays and links correctly.

**Comment by vermeeren** [comment]
This might be fixed with #512, released in Breathe v4.17.0. Note that you also need Sphinx 3.x for recent Breathe versions. Could someone try and post results?

**Comment by utzig** [comment]
> This might be fixed with #512, released in Breathe v4.17.0. Note that you also need Sphinx 3.x for recent Breathe versions. Could someone try and post results?

Yes we're using it and it works now.

**Comment by vermeeren** [comment]
@utzig Thanks for the update.
